### PR TITLE
Added translation for 2D mesh topology

### PR DESCRIPTION
**Pull request by michael-isaev**

### CODE REVIEW

File: paragraph/translation/allgather/mesh_1d_allgather_translator.cc [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            Opcode::kSend,
+            absl::StrCat(name_prefix,
+                         "_mesh-1d_ccw_send_",
+                         i),
+            allgather_sub_ptr));
+        ccw_sendrecv->AppendCommunicationGroup(
+            {comm_group.at(ccw_peer_index)});
+        ccw_sendrecv->SetBytesOut(comm_size / comm_group.size());
+      } else if (!ccw_send_flag && ccw_recv_flag) {
+      // If only receives from CCW but doesn't send there
```
Comment (nicmcd) [inline paragraph/translation/allgather/mesh_1d_allgather_translator.cc]: Indent problem

File: paragraph/translation/utils.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+#include <vector>
+
+#include "paragraph/graph/graph.h"
+
+namespace paragraph {
+
+// Helper functions for conversion bettwen consecutive processor id and
+// processor coordinates on a Mesh or Torus
+std::vector<uint64_t> ConsecutiveProcessorIdToMeshCoordinates(
```
Comment (nicmcd) [inline paragraph/translation/utils.h]: Call this either a cube (supersim does this) or a lattice (the proper term is an integer lattice).

File: paragraph/translation/utils.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    int64_t processor_id,
+    const std::vector<uint64_t>& dimension_sizes,
+    uint64_t concentration);
+
+uint64_t MeshCoordinatesToConsecutiveProcessorId(
+    const std::vector<uint64_t>& coordinates,
+    const std::vector<uint64_t>& dimension_sizes,
+    uint64_t concentration);
+
+// 2D swizzling that produces Hamiltonian cycle for all the verteces of 2D
```
Comment (nicmcd) [inline paragraph/translation/utils.h]: s/verteces/vertices

File: paragraph/translation/utils.cc [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  CommunicationGroup swizzled_ring;
+  for (uint64_t conc = 0; conc < concentration; conc++) {
+    swizzled_ring.push_back(
+        MeshCoordinatesToConsecutiveProcessorId({conc, 0, 0},
+                                                dimension_sizes,
+                                                concentration));
+  }
+  // Iterate over all columns; and all rows from the 1st to the last
+  for (uint64_t col = 0; col < dimension_sizes.at(0); col++) {
+    for (uint64_t j = 1; j < dimension_sizes.at(1); j++) {
```
Comment (nicmcd) [inline paragraph/translation/utils.cc]: What is "j"? Do you have a name that would work better?

File: paragraph/translation/utils.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+// Mesh or Torus. It is used to map these topologies to logical ring topology.
+CommunicationGroup Swizzling2dMeshToRing(
```
Comment (nicmcd) [inline paragraph/translation/utils.h]: Note: this works for mesh and torus but not all integer lattices. While it works for a torus, you are basically just ignoring some torus links. I'd name it Mesh in this case, mention it works for Torus too. However in the future you might have a specific Torus version as well.

What does swizzle mean? Can you say more about what the function does and returns? I'm guessing that you are walking the mesh and producing a list of IDs to be used as a ring. The word swizzle doesn't seem to work. I would suggest some like "CreateRingFromMesh".

File: paragraph/translation/allgather/mesh_2d_allgather_translator.cc [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  concentration_ = 1;
+  if (config.find("concentration") != config.end()) {
+    concentration_ = config["concentration"].get<uint64_t>();
+  }
+
+  // Create json config for internal 1D Mesh all-gather
+  nlohmann::json implicit_config = R"(
+    { "algorithm": "mesh-1d" }
+  )"_json;
+  // If we have a barrier in 2D Mesh, we need to instanciate a barrier before
```
Comment (nicmcd) [inline paragraph/translation/allgather/mesh_2d_allgather_translator.cc]: Typo, s/instanciate/instantiate, 
grep for this in other places

**Review by nicmcd — CHANGES_REQUESTED** [review]

File: paragraph/translation/allgather/mesh_1d_allgather_translator.cc (hunk 1 above)
Comment (michael-isaev) [reply paragraph/translation/allgather/mesh_1d_allgather_translator.cc]: fixed

File: paragraph/translation/utils.h (hunk 2 above)
Comment (michael-isaev) [reply paragraph/translation/utils.h]: Renamed to grid as it is less confusing (Cube can be confused with HyperCube, Integer Lattice kinda long and looks synonymous to grid) in PR#24

File: paragraph/translation/utils.h (hunk 3 above)
Comment (michael-isaev) [reply paragraph/translation/utils.h]: Fixed in PR#24

File: paragraph/translation/utils.cc (hunk 4 above)
Comment (michael-isaev) [reply paragraph/translation/utils.cc]: Fixed

File: paragraph/translation/allgather/mesh_2d_allgather_translator.cc (hunk 6 above)
Comment (michael-isaev) [reply paragraph/translation/allgather/mesh_2d_allgather_translator.cc]: Fixed

File: paragraph/translation/utils.h (hunk 5 above)
Comment (michael-isaev) [reply paragraph/translation/utils.h]: It is kinda similar to memory address swizzling that's common in GPUs. I updated comment with algorithm description and an example in PR#24

**Pull request merged**